### ISSUE TITLE
Add Automatic Concurrency Controls for Rewrite Datafiles Action

### ISSUE THREAD

**Issue by RussellSpitzer** [apache/iceberg#5140]
Currently we can only control the concurrency of a RewriteDatafiles Action by setting a static option, 

> By default the actions are executed serially, but can be run concurrently by increasing the value of **max-concurrent-file-group-rewrites**. This parameter controls the number of actions which will be run simultaneously.

This makes it difficult to properly set the parameter when some partitions require multiple tasks to complete and others require only a single task. For example a FileGroup which ends up writing 10 files will require 10 tasks and 10 Spark Cores. Another File-group may only have a single output file requiring only a single task.

Instead of using a static option, I think we should provide the option of allowing the Action to attempt to determine the number of open cores and schedule new jobs accordingly. This "auto" option would basically implement the following logic:

```
while (fileGroupsLeft is not empty) {
  for (group in group) {
    if (group.tasks > coresFree) {
      schedule(group)
    }
  }
  If (unable to schedule any groups and allCoresFree) {
    schedule(largestGroup)
  }
}
```
Basically we just always schedule jobs when we have open cores, if we have jobs that are all too large , then just schedule the largest.

**Comment by Zhangg7723** [comment]
Yes，the current design is coarse-grained, we are working on this point, I can do something for this issue.

**Comment by RussellSpitzer** [comment]
Glad to hear it, Let us know what you plan to do or if you have a PR for review

**Comment by ricardopereira33** [comment]
Hi @Zhangg7723!

Do you have any updates on this issue?
